[ONNX/Caffe2] Align Caffe2 Slice op with ONNX slice op using extended Caffe2 Slice op

Fiery opened the pull request:
ONNX's Slice supports axes as optional argument, however, this argument is not supported by Caffe2's Slice. So currently during model export, Caffe2 will turn the ONNX's Slice to 6 Caffe2 operators: Slice + ScatterAssign + Constant + ... , which may significantly decrease the performance.
This PR is to resolve this issue by extending Caffe2 Slice op to support an optional axes argument then eliminate redundant operators in exported Caffe2 Slice op from ONNX backend.

Commit:
[Caffe2][Slice] Extend Caffe2 Slice to support axes argument (#32737)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32737

Extends caffe2 slice operator to support an optional axes argument.

Differential Revision: D19544031

fbshipit-source-id: 39d4f5df4f609e0b71bca05c18aff69f584ab2a3

Commit:
[Caffe2][Slice] Extend Caffe2 Slice to support axes argument (#32737)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32737

Extends caffe2 slice operator to support an optional axes argument.

Differential Revision: D19544031

fbshipit-source-id: 39d4f5df4f609e0b71bca05c18aff69f584ab2a3

Commit:
[ONNX/Caffe2] Align Caffe2 Slice op with ONNX Slice during ONNX backend conversion using extended axes argument in Caffe2 Slice operator (#32725)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32725

Aligned Caffe2 Slice operator with ONNX Slice operator, reducing previously 6+ ops to 1 single op in converted Caffe2, with the help of newly extended axes argument in the Caffe2 Slice op.

Differential Revision: D19588582

fbshipit-source-id: afb62870140ffbc95c4fba3d18d3b65ccb200cc5

Commit:
[ONNX/Caffe2] Align Caffe2 Slice op with ONNX Slice during ONNX backend conversion using extended axes argument in Caffe2 Slice operator (#32725)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32725

Aligned Caffe2 Slice operator with ONNX Slice operator, reducing previously 6+ ops to 1 single op in converted Caffe2, with the help of newly extended axes argument in the Caffe2 Slice op.

Differential Revision: D19588582

fbshipit-source-id: afb62870140ffbc95c4fba3d18d3b65ccb200cc5

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit fcbd6459:


**None of the build failures appear to be your fault**.

* **1/1** [broken upstream](https://dr.pytorch.org/master-timeline.html?count=1&sha1=9bab617b3e6c7e80b74cf9b091cf5150e3571064&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35) at merge base 9bab617b since Jan 29
  <details>
  <summary>
  Please rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch origin viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=fcbd6459d3efe58a60900682b3a6511fa8dc67e8).

### :construction: 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably **caused by upstream breakages**:
* [pytorch_linux_backward_compatibility_check_test](https://dr.pytorch.org/build-details.html?build_id=93312954) from Jan 29



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

mruberry committed:
Fixes moving after weight norm application (#32563)

Summary:
This PR updates how RNNs handle their "flat weights." In particular, it allows for only some flat weights to be "materialized" when apply is called, and it updates the flattening behavior to only apply if all flat weights are (1) materialized, (2) share a dtype and (3) are acceptable to cuDNN.

One test is modified and another created to test these changes. One practical effect of this change is that weight norm can be successfully applied to a module BEFORE that module is moved to an accelerator. Previously doing so would throw an error.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32563

Differential Revision: D19602725

Pulled By: mruberry

fbshipit-source-id: d8f9441d17815c8c9ba15b256d4be36f784a3cf9

mrshenli committed:
Remove GIL from RRefContext (#32807)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32807

After this commit, RRefContext no longer depends on pybind.

Test Plan: Imported from OSS

Differential Revision: D19636316

Pulled By: mrshenli

fbshipit-source-id: 88faa101c32e9019e979ae8e5da6706e49842726

ezyang committed:
Rename DispatchKey::UndefinedTensorId to Undefined (#32728)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32728

It doesn't have much to do with tensors anymore.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Test Plan: Imported from OSS

Differential Revision: D19628093

Pulled By: ezyang

fbshipit-source-id: 4d57111cdf44ba347bec8a32bb5b4b47a83c1eaf

ezyang committed:
Make DispatchKeyGuards accept DispatchKey::Undefined (#32729)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32729

When working on the vmap prototype I noticed that this was helpful
as it lets me easily initialize a no-op guard, if I need to do it
at constructor time (which I usually do, because the guards don't
have move constructors).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Test Plan: Imported from OSS

Differential Revision: D19628092

Pulled By: ezyang

fbshipit-source-id: d6259a3f70d287cdac2e4a5f3984e2880f19bdc2

ezyang committed:
Centralize addition of "always on" dispatch keys. (#32734)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32734

VariableTensorId is the only key with this treatment today,
but BackendSelect and CompoundOp are coming soon.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Test Plan: Imported from OSS

Differential Revision: D19628091

Pulled By: ezyang

fbshipit-source-id: 250753f90528fa282af7a18d8d2f7736382754bd

Commit:
Resubmit more code fakefp16 mapping unification (#32798)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32798

ATT

Test Plan: unittests

Reviewed By: amylittleyang

Differential Revision: D19632251

fbshipit-source-id: 670004050d67415bb24392f3520afa32b64ce740

gaurav1086 committed:
[torch] fd error check

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/32797

Differential Revision: D19642262

Pulled By: mrshenli

fbshipit-source-id: 1720812166dd583dca6d72cb7e24b65ec013a62b

Commit:
Updating submodules

Summary:
GitHub commits:

https://github.com/facebook/fbthrift/commit/73638a87956241a954f71419ef8bc224a2642d23
https://github.com/facebook/folly/commit/7a83deaa8359e347674851600c68c6a574cd79c5
https://github.com/pytorch/fbgemm/commit/969d173d117670fd50e2aef85c63087190d1fcf5

Test Plan: n/a

Reviewed By: wittgenst

fbshipit-source-id: 399ed7a972876727a6bfd1409667c735c406fef5

houseroad committed:
[PyTorch BC] Fix the ci (#32843)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32843

fix the ci by skipping aten::join

Test Plan: ci

Reviewed By: hl475

Differential Revision: D19650584

fbshipit-source-id: 4446eef568ded334217ff9205a795daffebe41a1

ptrblck committed:
Cudnn bn size fix (#32763)

Summary:
Should fix https://github.com/pytorch/pytorch/issues/29744 by falling back to native batch norm implementation, if cudnn cannot execute the provided shape.

Shape numbers were verified for cudnn 7.6.5.32 with tensor shapes:
```python
# for spatial bn
x = torch.Size([880801, 256, 5])
x = torch.Size([65535, 256, 5])
x = torch.Size([880801, 64, 4, 4])
x = torch.Size([65535, 64, 4, 4])

# for per-act bn
x = torch.Size([131070, 2048])
x = torch.Size([262136, 2048])
```
for `training()` and `eval()` mode using `torch.float32` and `torch.float16`.

I've increased the shape of our current smoke test to, but I can also add all use cases of the support matrix, if wanted.

CC ngimel
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32763

Differential Revision: D19644328

Pulled By: ngimel

fbshipit-source-id: c2151bf9fe6bac79b8cbc69cff517a4b0b3867aa

zasdfgbnm committed:
Remove needs_dynamic_casting from TensorIterator and move it to Loops.cuh (#32755)

Summary:
Remove `needs_dynamic_casting` from TensorIterator and move it to `Loops.cuh`.

The original design of `needs_dynamic_casting` is fundamentally flawed: it injects logics into TensorIterator and uses a bunch of boolean values to test whether the dynamic casting is needed. This makes it very fragile, as the TensorIterator is so complicated and it is easy to introduce unnecessary dynamic casts. It also makes the `gpu_kernel` very unflexible, differently cases needs to manipulate TensorIterator to make it work.

For example, currently
```python
torch.zeros(10, device='cuda').mul_(0.9)
```
needs dynamic cast, but it shouldn't.

Testing whether dynamic casting is needed could be easy: just compare the dtypes of the lambda with the dtypes of operands. If they don't match, then dynamically cast, otherwise don't cast.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32755

Differential Revision: D19644092

Pulled By: ngimel

fbshipit-source-id: 130bb8bd78d20c2ed1bdfc9d9fb451eb0f0c7e55

hovhannesgithub committed:
Update Docs for building PyTorch for Android.

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/32578

Reviewed By: ljk53

Differential Revision: D19588904

Pulled By: dreiss

fbshipit-source-id: 2934752b9c5b94f2f141417669d8385be44d703b

Commit:
make tests for empty inputs check zero parameter grads (#32820)

Summary:
Make batch norm with empty inputs return zero parameter gradients. Now batch norm, group norm and convolutions now return zero grads for parameters, so make tests check that. Fixes some bullet points in https://github.com/pytorch/pytorch/issues/12013 (interpolate is not fixed by this PR, is being fixed in other PRs)
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32820

Differential Revision: D19651470

Pulled By: ngimel

fbshipit-source-id: 96fdd085f9b0e98e91217dd2ac1f30f9c482b8be

xush6528 committed:
Remove Python dependency (toPyTuple/fromPyTuple, jitCompilationUnit, deserialize) in rref_impl.h/cpp (#32753)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32753

Functions to be bound as an Aten operator could not have Python dependency.

This is to refactor and remove Python dependency.
ghstack-source-id: 97485800

Test Plan:
```
buck test mode/dev-nosan //caffe2/test/distributed/rpc:rpc_fork -- test_script_functions_not_supported

buck build mode/dev-nosan //caffe2/test/distributed/rpc:rpc_fork

buck-out/gen/caffe2/test/distributed/rpc/rpc_fork\#binary.par -r test_script_functions_not_supported
```

```
buck test mode/dev-nosan //caffe2/test/distributed/rpc:dist_autograd_fork

buck build mode/dev-nosan //caffe2/test/distributed/rpc:dist_autograd_fork

buck-out/gen/caffe2/test/distributed/rpc/dist_autograd_fork\#binary.par -r test_backward_simple_script_call
```

Differential Revision: D5741675

fbshipit-source-id: 31ee60955be8d815d0773f3699e3ff2f1f9d8849

Fiery committed:
Merge branch 'export-D19588582' of github.com:Fiery/pytorch into export-D19588582
